### PR TITLE
Handle window cursor

### PR DESCRIPTION
**Pull request by spatocode**
### Description:
This PR creates a better way of handling the window cursor with respect to canvasobjects.

Fixes #(726)

### Checklist:

- [x] Tests included.
- [x] Lint and formatter run with no errors.
- [x] Tests all pass.

#### Where applicable:

- [ ] Public APIs match existing style.
- [ ] Any breaking changes have a deprecation path or have been discussed.

### CODE REVIEW

File: internal/driver/glfw/window.go [hunk 1]
```diff
@@ -507,10 +507,10 @@ func (w *window) mouseMoved(viewport *glfw.Window, xpos float64, ypos float64) {
 
 	cursor := defaultCursor
 	obj, pos := w.findObjectAtPositionMatching(w.canvas, w.mousePos, func(object fyne.CanvasObject) bool {
-		if wid, ok := object.(*widget.Entry); ok {
-			if !wid.Disabled() {
-				cursor = entryCursor
-			}
+		entry, entOk := object.(*widget.Entry)
+		_, selOk := object.(*widget.SelectEntry)
+		if (entOk && !entry.Disabled()) || selOk {
```
Comment (andydotxyz) [inline internal/driver/glfw/window.go]: This ignores the disabled state of the select entry

**Comment by andydotxyz** [comment]
I think tests here could have caught the error whereby disabled state was ignored.
I know there are no cursor tests yet, so I have attached a diff that puts them in place for you.
Perhaps combining this with additions for your patch will round out the test side as well?
[cursor-test.diff.txt](https://github.com/fyne-io/fyne/files/4335234/cursor-test.diff.txt)

**Comment by spatocode** [comment]
> Good catch of another dependency of an internal package from `widget`.
> These should be removed where found because they introduce several kinds of problems.
> For instance nobody can create a custom entry widget with this cursor support. This should be solved the right way, i.e. we have to think about it.
> 
> Currently I can see two solutions to this:
> 
>     1. introduce special interfaces for cursor adjustment (currently entries and links: `fyne.Entry`, `fyne.Hyperlink`) and leave the selection of the cursor in the internal package
> 
>     2. introduce a new interface for canvas objects that want a cursor change (on hover) (`fyne.Cursable` wink) with the method `HoverCursor() fyne.Cursor` which returns a Fyne constant which can be translated to the according glfw constant by the internal package
> 
> 
> I like idea 2 most.
> Any better ideas?

You're right. I've thought of this already. If crafted well, it could help solve #726 .

File: internal/driver/glfw/window.go (hunk 1 above)
Comment (spatocode) [reply internal/driver/glfw/window.go]: The disabled state was already ignored before now. Could it be better to exempt the disabled state? I think it makes sense.

**Comment by spatocode** [comment]
I will begin work on refactoring this pull request inline with the discussion and suggestion by @toaster .

File: internal/driver/glfw/window.go (hunk 1 above)
Comment (andydotxyz) [reply internal/driver/glfw/window.go]: I'm not sure what you mean - the test above checks the disabled state of Entry - it only sets the bar cursor if the entry is not disabled.
The same should surely hold true for the selectEntry?
It seems fairly standard to not indicate an edit cursor over items that cannot be edited.

File: internal/driver/glfw/window_test.go [hunk 2]
```diff
@@ -59,6 +59,34 @@ func TestGLDriver_CreateSplashWindow(t *testing.T) {
...
 
+func cursorForObject(obj fyne.CanvasObject) *glfw.Cursor {
+	if wid, ok := obj.(*widget.Entry); ok {
+		if !wid.Disabled() {
+			return entryCursor
+		}
+	} else if _, ok := obj.(*widget.Hyperlink); ok {
+		return hyperlinkCursor
+	} else if _, ok := obj.(*widget.SelectEntry); ok {
+		return entryCursor
```
Comment (andydotxyz) [inline internal/driver/glfw/window_test.go]: I think this is a perfect spot to check disabled state like entry above :)

File: internal/driver/glfw/window_test.go [hunk 3]
```diff
@@ -59,6 +59,34 @@ func TestGLDriver_CreateSplashWindow(t *testing.T) {
...
+
+	assert.Same(t, defaultCursor, cursorForObject(widget.NewLabel("")))
+
+	entry := widget.NewEntry()
+	assert.Same(t, entryCursor, cursorForObject(entry))
+	entry.Disable()
+	assert.Same(t, defaultCursor, cursorForObject(entry))
+
+	selectEntry := widget.NewSelectEntry([]string{"A", "B", "C"})
+	assert.Same(t, entryCursor, cursorForObject(selectEntry))
```
Comment (andydotxyz) [inline internal/driver/glfw/window_test.go]: I think this is a perfect spot to check disabled state like entry above :)

File: canvasobject.go [hunk 4]
```diff
@@ -79,3 +79,9 @@ type Focusable interface {
 type Shortcutable interface {
 	TypedShortcut(Shortcut)
 }
+
+// Cursorable describes any CanvasObject that needs a cursor change
+type Cursorable interface {
+	Cursor() Cursor
+	SetCursor(Cursor)
```
Comment (toaster) [inline canvasobject.go]: I'm with @andydotxyz here: don't add `SetCursor()`. I don't think there is a real use case. IMO the cursor depends on the object's purpose and the purpose of an object does not change without the object being changed.

File: widget/widget.go [hunk 5]
```diff
@@ -12,6 +12,7 @@ type BaseWidget struct {
 	size     fyne.Size
 	position fyne.Position
 	Hidden   bool
+	cursor	 fyne.Cursor
```
Comment (toaster) [inline widget/widget.go]: I don't think that every widget should implement `fyne.Cursor`. Only the entries and links should, for now. Every non-cursorable object uses the default cursor. This saves a lot of redundant tests.

File: internal/driver/glfw/window.go [hunk 6]
```diff
@@ -502,17 +494,31 @@ func (w *window) findObjectAtPositionMatching(canvas *glCanvas, mouse fyne.Posit
...
 	obj, pos := w.findObjectAtPositionMatching(w.canvas, w.mousePos, func(object fyne.CanvasObject) bool {
-		entry, entOk := object.(*widget.Entry)
-		_, selOk := object.(*widget.SelectEntry)
-		if (entOk && !entry.Disabled()) || selOk {
-			cursor = entryCursor
-		} else if _, ok := object.(*widget.Hyperlink); ok {
-			cursor = hyperlinkCursor
+		if cursorable, ok := object.(fyne.Cursorable); ok {
+			fyneCursor := cursorable.Cursor()
+			cursor = cursorToName(fyneCursor)
```
Comment (toaster) [inline internal/driver/glfw/window.go]: AFAICS this creates a new glfw default cursor on every call. I'm not sure if this is save with glfw memory usage (might leak) and at least new Go objects are allocated every time.
I propose to return `initCursors` and create a map for the (immutable) default cursors.

File: internal/driver/glfw/window_test.go [hunk 7]
```diff
@@ -59,32 +59,26 @@ func TestGLDriver_CreateSplashWindow(t *testing.T) {
...
-	assert.Same(t, entryCursor, cursorForObject(entry))
+	assert.Equal(t, fyne.TextCursor, cursorForObject(entry))
 	entry.Disable()
-	assert.Same(t, defaultCursor, cursorForObject(entry))
+	assert.Equal(t, fyne.TextCursor, cursorForObject(entry))
 
 	selectEntry := widget.NewSelectEntry([]string{"A", "B", "C"})
-	assert.Same(t, entryCursor, cursorForObject(selectEntry))
+	assert.Equal(t, fyne.TextCursor, cursorForObject(selectEntry))
 }
```
Comment (toaster) [inline internal/driver/glfw/window_test.go]: This test does not really test anything on glfw.window. Instead it tests the `Cursor()` implementation of a couple of widgets.
The `Cursor()` implementation should be tested in the object specific tests (like entry_test and so on).
The window test should test the following cases:
* applied cursor for a non-cursorable object
* applied cursor for a `fyne.Cursorable` object which is _not_ `fyne.Disableable`(I suggest a minimal test object instead of an existing wigdet to not introduce a dependency from `widget` for this)
* applied cursor for a cursorable object which _is_ `fyne.Disableable` and enabled (again use a minimal test object)
* applied cursor for a disabled cursorable object

**Comment by toaster** [comment]
N.B.: Please ensure to run the glfw tests locally. They aren't part of the CI run, yet.

**Comment by andydotxyz** [comment]
You might find our suggested git hooks useful, to check formatting and lint before commit: https://github.com/fyne-io/fyne/wiki/GitHooks

File: internal/driver/glfw/window_test.go (hunk 7 above)
Comment (spatocode) [reply internal/driver/glfw/window_test.go]: Am not sure i understood your suggestion but i hope the new test works fine?

File: internal/driver/glfw/window.go [hunk 8]
```diff
@@ -494,31 +504,24 @@ func (w *window) findObjectAtPositionMatching(canvas *glCanvas, mouse fyne.Posit
...
+				cursor = textCursor
+			} else if fyneCursor == desktop.PointerCursor {
+				cursor = pointerCursor
+			} else if fyneCursor == desktop.CrosshairCursor {
+				cursor = crosshairCursor
+			} else if fyneCursor == desktop.HResizeCursor {
+				cursor = hResizeCursor
+			} else if fyneCursor == desktop.VResizeCursor {
+				cursor = vResizeCursor
+			}
```
Comment (toaster) [inline internal/driver/glfw/window.go]: Why didn't you keep a (constant) map which is initialised in `initCursors()`? This would reduce the amount of global vars and also simplify this code. Map lookup might be slower but I don't believe that this would be an issue.

File: internal/driver/glfw/window_test.go [hunk 9]
```diff
@@ -59,26 +60,32 @@ func TestGLDriver_CreateSplashWindow(t *testing.T) {
...
-	entry := widget.NewEntry()
-	assert.Equal(t, fyne.TextCursor, cursorForObject(entry))
-	entry.Disable()
-	assert.Equal(t, fyne.TextCursor, cursorForObject(entry))
-
-	selectEntry := widget.NewSelectEntry([]string{"A", "B", "C"})
-	assert.Equal(t, fyne.TextCursor, cursorForObject(selectEntry))
+	assert.Same(t, textCursor, cursorForObject(widget.NewEntry()))
+	assert.Same(t, pointerCursor, cursorForObject(widget.NewHyperlink("Testing", u)))
+	assert.Same(t, defaultCursor, cursorForObject(widget.NewButton("Test", nil)))
```
Comment (toaster) [inline internal/driver/glfw/window_test.go]: You are still testing that the widgets have a `Cursor()` implementation and which cursor they want.
This should be tested in the widgets' tests!

Here you should test that mouse moving over a `CanvasObject` which implements `Cursorable` changes the actual cursor.

If you want we can do a Google Meet session to talk about this and develop the tests together. Just contact me on Slack for that.

File: internal/driver/glfw/window.go [hunk 10]
```diff
@@ -504,24 +504,29 @@ func (w *window) findObjectAtPositionMatching(canvas *glCanvas, mouse fyne.Posit
 	return driver.FindObjectAtPositionMatching(mouse, matches, canvas.Overlays().Top(), roots...)
 }
 
+func cursorToName(cursor desktop.Cursor) *glfw.Cursor {
+	var cursorMap = map[desktop.Cursor]*glfw.Cursor{
```
Comment (toaster) [inline internal/driver/glfw/window.go]: Doesn't this create a new map on every call?

Comment (spatocode) [reply internal/driver/glfw/window.go]: Yes, it does but that shouldn't be an issue. We need to reduce the amount of global variables.

Comment (toaster) [reply internal/driver/glfw/window.go]: Why?

File: widget/select_entry_test.go [hunk 11]
```diff
@@ -1,4 +1,4 @@
-package widget_test
+package widget
```
Comment (toaster) [inline widget/select_entry_test.go]: I'm very biased regarding this. Why did you move the tests into the `widget` package?

File: widget/select_entry.go [hunk 12]
```diff
@@ -57,3 +58,19 @@ func (e *SelectEntry) SetOptions(options []string) {
 	dropDownButton.SetIcon(theme.MenuDropDownIcon())
 	e.ActionItem = dropDownButton
 }
+
+type cursorableButton struct {
+	Button
+}
+
```
Comment (toaster) [inline widget/select_entry.go]: Why not implement `Cursorable` in `Button` instead?

File: internal/driver/glfw/window_test.go [hunk 13]
```diff
@@ -60,32 +60,23 @@ func TestGLDriver_CreateSplashWindow(t *testing.T) {
...
+	assert.Same(t, textCursor, w.cursor)
+
+	w.mouseMoved(w.viewport, 10, float64(h.Position().Y+10))
+	assert.Same(t, pointerCursor, w.cursor)
 
-	assert.Same(t, textCursor, cursorForObject(widget.NewEntry()))
-	assert.Same(t, pointerCursor, cursorForObject(widget.NewHyperlink("Testing", u)))
-	assert.Same(t, defaultCursor, cursorForObject(widget.NewButton("Test", nil)))
+	w.mouseMoved(w.viewport, 10, float64(b.Position().Y+10))
+	assert.Same(t, defaultCursor, w.cursor)
```
Comment (toaster) [inline internal/driver/glfw/window_test.go]: 👍

File: internal/driver/glfw/window.go (hunk 10 above)
Comment (andydotxyz) [reply internal/driver/glfw/window.go]: Yes, this could be called a lot and I don't see the harm in a single map global to the package that is set up in init()

File: widget/select_entry_test.go (hunk 11 above)
Comment (spatocode) [reply widget/select_entry_test.go]: All widgets and corresponding tests in this package has the same package name `widget`. So i don't see any reason of not maintaining the the same package name throughout all widget test. Moreover, it's better calling functions from the same package than making redundant imports. 

For the special cursorable button, we already discussed this on slack. SelectEntry dropdown button cursor is overriden by it's parent's (SelectEntry) cursor. This solves that issue.

File: widget/select_entry.go (hunk 12 above)
Comment (spatocode) [reply widget/select_entry.go]: IMHO, I don't think that's the best thing to do. By default, any non-cursorable widget has `fyne.DefaultCursor`. I see no need of allowing `Button` implement `Cursorable` for the sake of `SelectEntry`. If button implement `Cursorable`, it's cursor will still remain `fyne.DefaultCursor`. Does it really make sense?

Comment (toaster) [reply widget/select_entry.go]: > Does it really make sense?

IMO yes. The implementation of `Cursorable` means that a component has a specific wish regarding the cursor. Any component that does not implement `Cursorable` simply doesn't care about the cursor.

Comment (toaster) [reply widget/select_entry.go]: Plus, I see no need to implement a special button just for the sake of the cursor in the `SelectEntry` :).

Comment (toaster) [reply widget/select_entry.go]: The drop-down button should behave like every other button. That's why it should not be a special button.

File: widget/select_entry_test.go (hunk 11 above)
Comment (toaster) [reply widget/select_entry_test.go]: I consider testing a package via its exported API the better approach. The only exception from this are unit tests for complex internal methods. But these are rare in my experience.

So moving tests into the tested package just to be able to implement an unnecessary internal struct is a bad idea.

File: internal/driver/glfw/window.go [hunk 14]
```diff
@@ -29,22 +28,26 @@ const (
 )
 
 var (
-	defaultCursor, entryCursor, hyperlinkCursor *glfw.Cursor
-	initOnce                                    = &sync.Once{}
-	defaultTitle                                = "Fyne Application"
+	defaultCursor, textCursor, pointerCursor, crosshairCursor, hResizeCursor, vResizeCursor *glfw.Cursor
```
Comment (andydotxyz) [inline internal/driver/glfw/window.go]: can you move this to two lines please - otherwise it makes the following aligned items harder to read

File: widget/select_entry.go (hunk 12 above)
Comment (andydotxyz) [reply widget/select_entry.go]: I have thought this over more and agree with @toaster - all buttons should have the same cursor, so let's just put this on widget.Button :)

**Comment by stuartmscott** [comment]
Excited to see the progress here, SplitContainer (#789) could use this to set the cursor to HResize/VResize when the user hovers over the divider.

File: driver/desktop/cursor.go [hunk 15]
```diff
@@ -0,0 +1,24 @@
+package desktop
+
+// Cursor represents a standard fyne cursor
+type Cursor string
+
+const (
+	// DefaultCursor is the default cursor typically an arrow
+	DefaultCursor Cursor = "DefaultCursor"
```
Comment (andydotxyz) [inline driver/desktop/cursor.go:8]: It just occurred to me also that this would probably be more efficient as an enum... using string comparison for an internal-only lookup seems like we're burning cycles for no benefit.

File: internal/driver/glfw/window.go (hunk 14 above)
Comment (andydotxyz) [reply internal/driver/glfw/window.go:31]: actually this probably doesn't matter, feel free to ignore that preference

File: internal/driver/glfw/window.go (hunk 10 above)
Comment (andydotxyz) [reply internal/driver/glfw/window.go:509]: Also I have noticed this method name is strange. Maybe it should be "nativeCursor" or "lookupNativeCursor"?

**Comment by andydotxyz** [comment]
I'd love to see this land - if possible in time for the 1.2.4 release I am about to start preparing. It would enable fixing issues that developers have with custom widgets not getting the right cursor.
Can you let us know if you'll be able to address the comments @spatocode?
Thanks for the work and sorry if this puts on pressure, really I just want to know for planning purposes.

**Comment by andydotxyz** [comment]
Great work thanks, I'm happy - how about @toaster ?

**Comment by toaster** [comment]
Awesome!
